File: Graph Algorithm/c++/hasPath.cpp
#include<bits/stdc++.h>
using namespace std;

bool hasPathRecursive(map<string,vector<string>>graph, string source, string des)
{
    if (source==des) return true;

    vector<string>::iterator i;
    for (i=graph[source].begin();i!=graph[source].end();i++)
        if (hasPathRecursive(graph,*i,des))
            return true;

    return false;
}

bool hasPath(map<string,vector<string>>graph, string source, string des)
{
    queue<string>q;
    q.push(source);

    while (q.size()>0)
    {
        string current=q.front();q.pop();
        if (current==des) return true;

        vector<string>::iterator i;
        for (i=graph[current].begin();i!=graph[current].end();i++)
            q.push(*i);
    }

    return false;
}

int main()
{
    map<string,vector<string>>graph={{"f",{"g","i"}},{"g",{"h"}},{"h",{}},{"i",{"g","k"}},{"j",{"i"}},{"k",{}}};
    cout<<"test1: "<<hasPath(graph,"j","f")<<" - "<<hasPathRecursive(graph,"j","f")<<endl;
    cout<<"test2: "<<hasPath(graph,"h","g")<<" - "<<hasPathRecursive(graph,"h","g")<<endl;
    cout<<"test3: "<<hasPath(graph,"f","k")<<" - "<<hasPathRecursive(graph,"f","k")<<endl;
}

File: Graph Algorithm/python/hasPath.py
def hasPathRecursive(graph,source,des):
    if source==des: return True

    for i in graph[source]:
        if hasPathRecursive(graph,i,des): return True

    return False

def hasPath(graph,source,des):
    queue=[source]

    while len(queue)>0:
        current=queue.pop(0)
        if current==des: return True

        for i in graph[current]:
            queue.append(i)

    return False

graph={
    'f':['g','i'],
    'g':['h'],
    'h':[],
    'i':['g','k'],
    'j':['i'],
    'k':[]
}

print('test1',hasPath(graph,'f','k'),hasPathRecursive(graph,'f','k'))
print('test2',hasPath(graph,'j','f'),hasPathRecursive(graph,'j','f'))
print('test3',hasPath(graph,'h','g'),hasPathRecursive(graph,'h','g'))
